### ISSUE TITLE
[MSHADE-174] Unable to shade Java 8 jarfiles with static interface methods using minimizeJar

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-shade-plugin#529]
**[Pedro Boado](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rogue)** opened **[MSHADE-174](https://issues.apache.org/jira/browse/MSHADE-174?redirect=false)** and commented

Maven shade does not appear to work with shading in Java 8 jarfiles that include static interface methods. A full log of a build with dependency resolution is attached (as well as an effective-pom and the stacktrace)

The error header is:

```
Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:2.3:shade (default) on project --: Error creating shaded jar: INVOKESPECIAL/STATIC on interfaces require ASM 5 -> [Help 1]
```

(Do note, I already am using ASM 5.0.2)

I've also created a public, reproducible version of this bug: https://github.com/1Rogue/Maven-Shade-Bug

You simply can clone that repo and attempt a "mvn clean install", and it should fail accordingly.

You can also view the stackoverflow question that sent me here: http://stackoverflow.com/questions/24415677/unable-to-shade-jar-with-maven-invokespecial-static/24417179#24417179

---

**Affects:** 2.2, 2.3

**Attachments:**
- [diff.1805](https://issues.apache.org/jira/secure/attachment/12719020/diff.1805) (_1.79 kB_)
- [effective-pom.txt](https://issues.apache.org/jira/secure/attachment/12719027/effective-pom.txt) (_10.58 kB_)
- [out.txt](https://issues.apache.org/jira/secure/attachment/12718975/out.txt) (_9.76 kB_)
- [stacktrace.txt](https://issues.apache.org/jira/secure/attachment/12718927/stacktrace.txt) (_3.50 kB_)

**Issue Links:**
- [MSHADE-166](https://issues.apache.org/jira/browse/MSHADE-166) maven-shade-plugin not compatible with Java 8

1 votes, 3 watchers

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

I did most of the asm5 ports which were necessary to make anything work /at all/ with java8, and I have been waiting for this issue to  pop up :)

The thing is, there are probably some new asm api's that need to be used to iterate the new jdk8 features, and these have not been implemented anywhere AFAIK. So if you add a static interface method, and reference a type **only** from that method, the minimizeJar feature will not "see" that this type is in use.  Same problem might apply to lambdas (maybe)

The solution is to patch stuff so that the new asm methods are called and the types are added as "used". Patches are welcome and will be reviewed :)

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

Applied in e04f22b36268ec90019702f6df223acef954f27e on jdependency

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

Darn. Not any more. Did anyone see that :) ? git reset --hard origin/master if you did :)

You do not really need to recompile shade to test this fix, you can just add the 1.0-SNAPSHOT dependency to the plugin definition to override the old version.

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

Fixed in r1664986
